### PR TITLE
[SPARK-6963][CORE]Flaky test: o.a.s.ContextCleanerSuite automatically cleanup checkpoint

### PR DESCRIPTION
**Pull request by witgo**
cc @andrewor14

### CODE REVIEW

**Comment by srowen** [comment]
I'd like to start pushing back on PRs without explanation, here or in the JIRA. Please add a short explanation of why this is the right change and what the problem is. Otherwise I think we need to start rejecting these.

File: core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala [hunk 1]
```diff
@@ -245,7 +245,7 @@ class ContextCleanerSuite extends ContextCleanerSuiteBase {
     assert(fs.exists(RDDCheckpointData.rddCheckpointDataPath(sc, rddId).get))
 
     // Test that GC causes checkpoint data cleanup after dereferencing the RDD
-    postGCTester = new CleanerTester(sc, Seq(rddId), Nil, Nil)
+    postGCTester = new CleanerTester(sc, Seq(rddId), Nil, Nil, Seq(rddId))
```
Comment (witgo) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala]: @srowen This is just a test code bug. `new CleanerTester(sc, Seq(rddId), Nil, Nil)`  This code is only to ensure that the RDD is cleaned, but checkpoint.  rdd and checkpoint almost simultaneously be cleaned, But there are exceptions, depending on the GC.  The `checkpointCleaned ` ensure the checkpoint is cleaned.

Comment (srowen) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala]: How does this fix the assertion that failed? you add a new set of IDs that must now be counted down by a new callback, and you assert it ends up empty, which is good. But the assertion that failed was earlier:
```
     assert(fs.exists(RDDCheckpointData.rddCheckpointDataPath(sc, rddId).get))
```

Comment (witgo) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala]: The tips are more confusing. 
The [jira](https://issues.apache.org/jira/browse/SPARK-6963) display the error in: [ContextCleanerSuite.scala#L252](https://github.com/witgo/spark/blob/SPARK-6963/core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala#L252).

Comment (srowen) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala]: Ah right, it is the following one. Still, how does this affect whether the checkpoint files exist?

File: core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala [hunk 2]
```diff
@@ -427,12 +429,17 @@ class CleanerTester(
...
-      logInfo("Broadcast" + broadcastId + " cleaned")
+      logInfo("Broadcast " + broadcastId + " cleaned")
     }
 
     def accumCleaned(accId: Long): Unit = {
       logInfo("Cleaned accId " + accId + " cleaned")
     }
+
+    def checkpointCleaned(rddId: Long): Unit = {
+      toBeCheckpointIds -= rddId
```
Comment (witgo) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala]: @srowen  When the checkpoint is cleaned, `checkpointCleaned ` is called asynchronous,  `isAllCleanedUp` returns true,  and then `postGCTester.assertCleanup()`  return.

Comment (srowen) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala]: Great, that's a good explanation. I see that this makes it effectively wait longer to proceed, by which time some necessary conditions are true, like that checkpoint cleanup has happened. Thank you for that detail, and to the limits of my knowledge, this looks good.

**Pull request closed without merging**